components/script/dom/bindings/settings_stack.rs

```diff
@@ -0,0 +1,67 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use dom::bindings::js::{JS, Root};
+use dom::globalscope::GlobalScope;
+use std::cell::RefCell;
+use js::jsapi::JSTracer;
+use dom::bindings::trace::JSTraceable;
+
+thread_local!(static STACK: RefCell<Vec<StackEntry>> = RefCell::new(Vec::new()));
+
+#[allow(unrooted_must_root)]
+#[derive(JSTraceable)]
+struct StackEntry {
+    global: JS<GlobalScope>,
+}
+
+/// Traces the script settings stack.
+pub unsafe fn trace(tracer: *mut JSTracer) {
+    STACK.with(|stack| {
+        stack.borrow().trace(tracer);
+    })
+}
+
+/// RAII struct that pushes and pops entries from the script settings stack.
+pub struct AutoEntryScript {
+    _private: (),
+}
+
+impl AutoEntryScript {
+    /// https://html.spec.whatwg.org/multipage/#prepare-to-run-script
+    pub fn new(global: &GlobalScope) -> Self {
+        trace!("Prepare to run script with {:p}", global);
+        STACK.with(|stack| {
+            let mut stack = stack.borrow_mut();
+            stack.push(StackEntry {
+                global: JS::from_ref(global),
+            });
+            AutoEntryScript {
+                _private: (),
+            }
+        })
+    }
+}
+
+impl Drop for AutoEntryScript {
+    /// https://html.spec.whatwg.org/multipage/#clean-up-after-running-script
+    fn drop(&mut self) {
+        let entry = STACK.with(|stack| {
+            let mut stack = stack.borrow_mut();
+            stack.pop().unwrap()
```

Comment:
`AutoEntryScript` can move so these can be dropped out of order, I'm pretty sure that's not intended. Seems like we want something similar to `rooted!`, which would make the whole TLS stack useless AFAICT?